migration-engine/connectors/sql-migration-connector/src/sql_schema_differ/sql_schema_differ_flavour/mssql.rs

```diff
@@ -39,20 +42,1064 @@ impl SqlSchemaDifferFlavour for MssqlFlavour {
+
+            match (previous_type, next_type) {
+                (None, _) | (_, None) => family_comparison(),
+                (Some(previous), Some(next)) if previous == next => None,
+                (Some(previous), Some(next)) => Some(native_type_change_riskyness(previous, next)),
+            }
         } else {
-            Some(type_change_riskyness(differ))
+            family_comparison()
         }
     }
 }
 
-fn type_change_riskyness(differ: &ColumnDiffer<'_>) -> ColumnTypeChange {
-    match (differ.previous.column_type_family(), differ.next.column_type_family()) {
+fn family_change_riskyness(previous: &ColumnTypeFamily, next: &ColumnTypeFamily) -> ColumnTypeChange {
+    match (previous, next) {
         (_, ColumnTypeFamily::String) => ColumnTypeChange::SafeCast,
         (ColumnTypeFamily::String, ColumnTypeFamily::Int)
         | (ColumnTypeFamily::DateTime, ColumnTypeFamily::Float)
         | (ColumnTypeFamily::String, ColumnTypeFamily::Float) => ColumnTypeChange::NotCastable,
         (_, _) => ColumnTypeChange::RiskyCast,
     }
 }
+
+fn native_type_change_riskyness(previous: MsSqlType, next: MsSqlType) -> ColumnTypeChange {
+    use ColumnTypeChange::*;
+    use MsSqlTypeParameter::*;
+
+    println!("{:?} -> {:?}", previous, next);
+
+    match previous {
+        // Bit, as in booleans. 1 or 0.
+        MsSqlType::Bit => match next {
+            MsSqlType::TinyInt => SafeCast,
+            MsSqlType::SmallInt => SafeCast,
+            MsSqlType::Int => SafeCast,
+            MsSqlType::BigInt => SafeCast,
+            MsSqlType::Decimal(_) => SafeCast,
+            MsSqlType::Numeric(_) => SafeCast,
+            MsSqlType::Money => SafeCast,
+            MsSqlType::SmallMoney => SafeCast,
+            MsSqlType::Float(_) => SafeCast,
+            MsSqlType::Real => SafeCast,
+            MsSqlType::DateTime => SafeCast,
+            MsSqlType::SmallDateTime => SafeCast,
+            MsSqlType::Binary(_) => SafeCast,
+            MsSqlType::VarBinary(_) => SafeCast,
+            MsSqlType::Bit => SafeCast,
+            MsSqlType::Char(_) => SafeCast,
+            MsSqlType::NChar(_) => SafeCast,
+            MsSqlType::VarChar(_) => SafeCast,
+            MsSqlType::NVarChar(_) => SafeCast,
+            _ => NotCastable,
+        },
+
+        // Maps to u8.
+        MsSqlType::TinyInt => match next {
+            MsSqlType::TinyInt => SafeCast,
+            MsSqlType::SmallInt => SafeCast,
+            MsSqlType::Int => SafeCast,
+            MsSqlType::BigInt => SafeCast,
+            MsSqlType::Decimal(params) => match params {
+                // TinyInt can be at most three digits, so this might fail.
+                Some((p, s)) if p - s < 3 => RiskyCast,
+                _ => SafeCast,
+            },
+            MsSqlType::Numeric(params) => match params {
+                // TinyInt can be at most three digits, so this might fail.
+                Some((p, s)) if p - s < 3 => RiskyCast,
```

Comment:
i assume p stands for precision and s for size: wouldn't the correct thing be `size - precision`? (precision-size should be negative). Not sure my mental model is right, here.

Reply:
Not really, `p` is precision and `s` is scale. `p` means how many digits we can have, `s` how many digits we have after the comma on the right side.

Reply:
So `Decimal(4, 2)` means our numbers look like `12.12`. `Decimal(4, 0)` would look like `1212`.

Reply:
ok, makes sense